grpc/client.go

```diff
@@ -0,0 +1,363 @@
+
+	// Retry represents the retry configuration
+	Retry *RetryConfig `mapstructure:"Retry"`
+}
+
+// DefaultConfig returns a default configuration for the gRPC client
+func DefaultConfig() *ClientConfig {
+	return &ClientConfig{
+		URL:               "localhost:50051",
+		MinConnectTimeout: types.NewDuration(defaultTimeout),
+		Retry: &RetryConfig{
+			MaxAttempts:       defaultMaxAttempts,
+			InitialBackoff:    types.NewDuration(defaultInitialBackoff),
+			MaxBackoff:        types.NewDuration(defaultMaxBackoff),
+			BackoffMultiplier: defaultBackoffMultiplier,
+		},
+		RequestTimeout: types.NewDuration(defaultTimeout),
+		UseTLS:         false,
+	}
+}
+
+// String returns a string representation of the gRPC client configuration
+func (c *ClientConfig) String() string {
+	if c == nil {
+		return noneStr
+	}
+
+	return fmt.Sprintf("GRPC Client Config: "+
+		"URL=%s, MinConnectTimeout=%s, "+
+		"RequestTimeout=%s, UseTLS=%t, Retry=%s",
+		c.URL, c.MinConnectTimeout.String(),
+		c.RequestTimeout.Duration, c.UseTLS, c.Retry.String())
+}
+
+// Validate checks if the gRPC client configuration is valid.
+// It returns an error if any of the required fields are missing or invalid.
+func (c *ClientConfig) Validate() error {
+	if c == nil {
+		return fmt.Errorf("gRPC client configuration cannot be nil")
+	}
+
+	if c.URL == "" {
+		return fmt.Errorf("gRPC client URL cannot be empty")
+	}
+
+	if c.MinConnectTimeout.Duration <= 0 {
+		return fmt.Errorf("MinConnectTimeout must be greater than zero")
+	}
+
+	if c.Retry != nil {
+		if err := c.Retry.Validate(); err != nil {
+			return err
+		}
+
+		initialBackoffMillis := float64(c.Retry.InitialBackoff.Milliseconds())
+		attempts := float64(c.Retry.MaxAttempts)
+
+		// minTimeout = initialBackoff - (1 - backoffMultiplier ^ maxAttempts) / (1 - backoffMultiplier)
+		minTimeoutMillis := initialBackoffMillis *
+			(1 - math.Pow(c.Retry.BackoffMultiplier, attempts)) / (1 - c.Retry.BackoffMultiplier)
+
+		minRequestTimeout := time.Duration(minTimeoutMillis * float64(time.Millisecond))
+		if c.RequestTimeout.Duration < minRequestTimeout {
+			return fmt.Errorf("RequestTimeout (%s) is too short; expected at least %s to accommodate retries",
+				c.RequestTimeout, minRequestTimeout)
+		}
+	}
+
+	return nil
+}
+
+// RetryConfig denotes the gRPC retry policy
+type RetryConfig struct {
+	// InitialBackoff is the initial delay before retrying a request
+	InitialBackoff types.Duration `mapstructure:"InitialBackoff"`
+
+	// MaxBackoff is the maximum backoff duration for retries
+	MaxBackoff types.Duration `mapstructure:"MaxBackoff"`
+
+	// BackoffMultiplier is the multiplier for the backoff duration
+	BackoffMultiplier float64 `mapstructure:"BackoffMultiplier"`
+
+	// MaxAttempts is the maximum number of retries for a request
+	MaxAttempts int `mapstructure:"MaxAttempts"`
+
+	// Excluded captures functions which are excluded from retry policies
+	Excluded []Method `mapstructure:"Excluded"`
+}
+
+type Method struct {
```

Comment:
Move this struct after `RetryConfig` logic.

Reply:
Fixed https://github.com/agglayer/aggkit/pull/524/commits/33643b1e14622bfce3384fc5ee68bd764d16365f